[로니] 스프링 카페 2단계 - 글 쓰기 기능 구현

CMSSKKK opened the pull request:
안녕하세요! 로니입니다😄

2단계 PR 보냅니다!

## 구현

- [x] 글쓰기
- [x] 글 목록 조회하기
- [x] 게시글 상세보기
- [x] 회원정보 수정

항상 꼼꼼하게 리뷰해주시고, 공부할 키워드와 생각할 거리를 남겨주셔서 감사합니다!!

CMSSKKK committed:
refactor : 1단계 리뷰 내용 검토 및 수정

1. 접근 제어자 수정
2. User 객체 setter 삭제, User userDto를 파라미터로 받는 생성자 생성
3. UserService findUser orElseThrow 사용
4. controllerTest displayname 수정

CMSSKKK committed:
feat : validation dependency 추가 및 UserDto validation 어노테이션 추가

CMSSKKK committed:
Test : UserDtoTest validation 어노테이션 적용 테스트

CMSSKKK committed:
refactor : user default 생성자 삭제 및 변수 final로 수정

CMSSKKK committed:
feat : UserException 발생시 error-page/400으로 보내기

CMSSKKK committed:
rename : templates user디렉토리 생성 및 qna디렉토리 이동 및 중복코드 제거

CMSSKKK committed:
feat : Post 객체, PostRepository, MemoryPostsRepository 구현

CMSSKKK committed:
refactor : UserException ClientException으로 변경 및 ResponseStatus 어노테이션 추가

CMSSKKK committed:
feat : ArticleService 구현 및 Posts -> Article rename

CMSSKKK committed:
test : MemoryArticleRepository ArticleService 테스트

CMSSKKK committed:
refactor : checkIndex() 메서드를 통한 findOne() 예외처리, 변수명 rename

CMSSKKK committed:
feat : ArticleController 구현 view 매핑

ArticleResponseDto 구현, 기존 ArticleDto로 반환하는 것을 ArticleResponseDto로 반환하도록 변경

CMSSKKK committed:
test : ArticleController webMvcTest 작성

CMSSKKK committed:
feat : 회원정보 수정 기능 구현 및 webLayerTest추가

CMSSKKK committed:
test : updateUserInfo test

CMSSKKK committed:
rename : user/create -> user/sign-up, UserService.findById -> findOne 으로 수정

tmdgusya commented on src/main/java/com/kakao/cafe/domain/article/Article.java:
```diff
@@ -0,0 +1,57 @@
+        if (o == null || getClass() != o.getClass()) return false;
+        Article article = (Article) o;
+        return id == article.id && Objects.equals(writer, article.writer) && Objects.equals(title, article.title) && Objects.equals(contents, article.contents);
+    }
```
전부 같아야만 같은 값인가요? 다름을 구별할 수 있는 유니크한 값은 없나요?

CMSSKKK replied:
Value Object와 Entity의 차이에 대해서 그룹원이 좋은 글을 공유해준 덕분에 더 고민할 수 있었습니다. 
https://www.culttt.com/2014/04/30/difference-entities-value-objects/

로치가 코멘트남겨주신 의도도 이와 같을까요..?

위 글을 토대로 생각을 정리했을 때, `Article` 객체를 entity로 볼 수 있을 것 같습니다. 

entity의 경우라면 `id`와 같은 identifier만을 체크해서 객체를 판별하면 될까요?

tmdgusya replied:
아뇨 제 뜻은 그냥 단순히 entity / value object 가 아니라 유니크한 부분이 없는지 물어본거에요.
unique 한 부분이 있으면 그 부분만 비교해도 되지 않을까 해서요

tmdgusya commented on src/main/java/com/kakao/cafe/domain/article/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,48 @@
+
+    private static final int STORAGE_KEY = 1;
+
+    private final List<Article> articles = new ArrayList<>();
```
자료구조로 List 를 고르신 이유는 무엇일까요?

CMSSKKK replied:
요구사항이라 고민없이 구현한 것 같습니다ㅠㅠ 

요구사항에 의존하기보다 효율적인 자료구조에 대해서 고민하고 구현하겠습니다!

tmdgusya commented on src/main/java/com/kakao/cafe/domain/article/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public void save(Article article) {
+        article.setId(generateId());
```
자료구조로 List 를 사용하였음에도 불구하고 id 가 필요한 이유는 무엇인가요?

CMSSKKK replied:
article을 구별할 수 있는 고유한 값이 필요하다고 생각합니다!

Map을 활용하도록 해보겠습니다.

tmdgusya commented on src/main/java/com/kakao/cafe/domain/article/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,48 @@
+    @Override
+    public Optional<Article> findById(int id) {
+        int index = id - STORAGE_KEY;
+        if(checkIndex(index)) {
```
`checkIndex` 라는 건 정확하게 어떤걸 체크하는지 알기 힘들어보이기도 하네요 ㅎㅎ

CMSSKKK replied:
조금 더 고민해보겠습니다!ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/domain/user/User.java:
```diff
@@ -16,10 +16,21 @@ public User(String userId, String password, String name, String email) {
         this.email = email;
     }
 
+    public User(UserDto userDto) {
+        this.userId = userDto.getUserId();
+        this.password = userDto.getPassword();
+        this.name = userDto.getName();
+        this.email = userDto.getEmail();
+    }
```
보통 이럴경우 생성자로 해도 되지만 `of` 혹은 `from` 을 많이 사용해요.
한번 이 부분도 찾아보시고 적용해보셔도 좋을거 같아요 ㅎㅎ

CMSSKKK replied:
저번에 코멘트해주셔서 공부를 했었는데 활용을 못했네요 😂 

활용해보겠습니다 감사합니다!

tmdgusya commented on src/main/java/com/kakao/cafe/domain/user/User.java:
```diff
@@ -36,19 +47,4 @@ public String getEmail() {
-
-    public void setEmail(String email) {
-        this.email = email;
-    }
```
좋습니다 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,42 @@
+
+    public ArticleResponseDto findOne(int id) {
+        Article article = articleRepository.findById(id).orElseThrow(() -> {
+            throw new ClientException(HttpStatus.BAD_REQUEST, "게시글을 찾을 수 없습니다.");
```
BAD REQUEST 가 적절할까요? 더 적절한게 있을거 같은데

Github 에 주소창에 아무 Repository 나 입력하시고 없을경우 어떤 Status Code 가 나오는지 확인해보셔도 좋을거 같아요

tmdgusya replied:
HTTP METHOD 에 대해서도 공부해 보시는걸 추천드립니다.

CMSSKKK replied:
예외가 구현의 우선순위가 아니라 생각해서 깊은 고민이 없었던 것 같습니다!

항상 꼼꼼하게 공부하겠습니다. 키워드 감사합니다~

tmdgusya commented on src/main/java/com/kakao/cafe/service/UserService.java:
```diff
@@ -32,20 +33,36 @@ public void join(User user) {
         });
     }
+
+    private User findOne(String userId) {
```
굳이 이 부분을 따로 빼낼이유는 없어보이긴 하는데요 ㅎㅎ
뭐 이건 취향이니 안고치셔도 상관은 없습니다.

CMSSKKK replied:
코드가 중복되서 따로 빼봤습니다.. ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/service/UserService.java:
```diff
@@ -32,20 +33,36 @@ public void join(User user) {
     private void checkDuplicateId(User user) {
         userRepository.findById(user.getUserId()).ifPresent(u -> {
-            throw new IllegalArgumentException("중복된 아이디가 있습니다.");
+            throw new ClientException(HttpStatus.BAD_REQUEST, "아이디가 이미 존재합니다.");
```
여기도 마찬가지로 BAD_REQUEST 가 적절할까요 ㅎㅎ?
한번 고민해보시죠.

CMSSKKK replied:
공부하고 수정하겠습니다!

tmdgusya commented on src/main/java/com/kakao/cafe/web/ArticleController.java:
```diff
@@ -0,0 +1,55 @@
+import org.springframework.web.bind.annotation.PostMapping;
+
+@Controller
+public class ArticleController {
```
`/qna` 가 중복되는거 같은데 Class 상단에 `@RequestMapping` 을 붙이면 중복제거가 가능합니다.
한번 공부해보시고 적용해보세요 ㅎㅎ

CMSSKKK replied:
넵 중복제거하겠습니다 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/web/ArticleController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @PostMapping("/qna/write-qna")
+    public String write(ArticleDto articleDto) {
```
어떻게 아무것도 안붙였는데 Request Body 를 Dto 로 전환해줄까요 ㅎㅎ? 이 부분에 대해서 공부해보셨나요?

CMSSKKK replied:
어노테이션이 없는 경우 `@RequestParam` 혹은 `@ModelAttribute` 으로 처리하는 것으로 알고 있습니다!

 `@ModelAttribute`는 생성자를 통해서 필드를 찾아서 매핑하는 것으로 알고 있습니다!

tmdgusya replied:
좋네요 :)

tmdgusya commented on src/main/java/com/kakao/cafe/web/UserController.java:
```diff
@@ -47,16 +49,33 @@ public String showUsersAll(Model model) {
-        UserResponseDto userResponseDto = userService.findUser(userid);
+    @GetMapping("/users/{id}")
+    public String showProfile(@PathVariable String id, Model model) {
+        logger.info("search {}", id);
```
로그를 좀만 더 자세하게 찍어보시죠 ㅎㅎ

CMSSKKK replied:
로그에 대해서도 더 공부하고, 고민하겠습니다! 

로그도 네이밍과 같군요 ....ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/web/dto/ArticleDto.java:
```diff
@@ -0,0 +1,60 @@
+
+public class ArticleDto {
+
+    @NotBlank
```
왜 Dto 에서 유효성 검증을 해야 할까요? 이 부분에 대해서 고민하거나 공부해보신적 있으신가요?

CMSSKKK replied:
계층 분리에 대해서 많이 찾아보고 슬랙에 공유해주신 로치의 블로그 글도 봤습니다 ㅎㅎ 

아직 완벽하게 정리하고 제 지식이 되지는 못했지만, 
`Article`을 entity로 보고 다른 책임을 갖지 않도록 하자라는 생각으로 dto에서 검증하도록 했습니다.

tmdgusya replied:
좋네요 ㅎㅎ

tmdgusya commented on src/test/java/com/kakao/cafe/web/dto/UserDtoTest.java:
```diff
@@ -0,0 +1,74 @@
+        UserDto userDtoEmpty = new UserDto(" ", " ", " ", " ");
+        Set<ConstraintViolation<UserDto>> violations = validator.validate(userDtoEmpty);
+
+        violations.forEach(i -> System.out.println(i.getMessage()));
```
그다지 좋아보이는 문장은 아닌거 같은데요 ㅎㅎ? 
테스트에서 로그를 찍어야 하나 라는 의문이 있습니다.

CMSSKKK replied:
Validation 어노테이션을 처음 사용하면서 예제와 비슷하게 테스트를 구현하고, message를 확인해보고자 남긴 코드입니다...

수정하겠습니다!

tmdgusya reviewed (changes requested):
전반적으로 자료구조에 `final` 을 사용하신 부분이 좋네요.
코드가 깔끔한 부분들이 많습니다.

아쉬운점은 로그를 찍는 부분이 약간 부실한거 같습니다. 
혹시 코드를 안보고 저 로그만 보고 에러를 찾거나, 어떤 일이 일어났는지 추측이 가능하신가요?
로그는 추적하기 쉬우라고 찍는 것이므로 잘 전달될 수 있는 정보들을 기입해야 합니다.
조금 만 더 신중하게 작성해보시죠 :)

다만 궁금한점과 수정할 사항들이 있어 리뷰 남겨드렸습니다.

CMSSKKK committed:
refactor : request changes 수정 반영

User, Article equals 메서드 수정
checkIndex() -> hasIndexInArticlesBound()
User 생성자 정적 팩토리 메서드 of로 수정
예외  HttpStatus 수정
ArticleController @RequestMapping 추가
로그 내용 수정
UserDtoTest 로그 프린트 삭제

CMSSKKK commented:
안녕하세요 로치😃

코멘트 남겨주신 부분 조금 더 고민하고 수정했습니다.

List 자료구조는 3단계에서 db를 연결하면서 같이 수정해보려고 합니다.
또한 로그에 대해서도 조금 더 공부하고 더 수정해보겠습니다.

코드를 작성하면서 고민하고 해결하는 노력을 하지만, 아직 많이 부족한 것 같습니다.

항상 꼼꼼한 코멘트 덕분에 원리도 조금 더 공부하고, 좋은 코딩 습관을 가질 수 있도록 더 연습하게 된 것 같습니다.

휴일 잘 보내시고 항상 건강 잘 챙기셨으면 좋겠습니다! 감사합니다👍

tmdgusya commented:
고생하셨습니다 :)
